modules/redis/src/main/scala/zio/redis/SingleNodeSubscriptionExecutor.scala

```diff
@@ -0,0 +1,219 @@
+    ) = for {
+      subscription <- subscriptionRef.get
+      _ <- ZIO.foreachDiscard(subscription.get(key))(
+             ZIO.foreachDiscard(_)(queue => queue.offer(Take.single(msg)).unlessZIO(queue.isShutdown))
+           )
+    } yield ()
+
+    def releaseStream(subscriptionRef: Ref[Map[String, Chunk[Queue[Take[RedisError, PushProtocol]]]]], key: String) =
+      for {
+        subscription <- subscriptionRef.getAndUpdate(_ - key)
+        _ <- ZIO.foreachDiscard(subscription.get(key))(
+               ZIO.foreachDiscard(_)(queue => queue.offer(Take.end).unlessZIO(queue.isShutdown))
+             )
+      } yield ()
+
+    connection.read
+      .mapError(RedisError.IOError(_))
+      .via(RespValue.Decoder)
+      .collectSome
+      .mapZIO(resp => ZIO.attempt(PushProtocolOutput.unsafeDecode(resp)))
+      .refineToOrDie[RedisError]
+      .foreach {
+        case msg @ PushProtocol.Subscribe(channel, _) => offerMessage(channelSubsRef, channel, msg)
+        case msg @ PushProtocol.Unsubscribe(channel, _) =>
+          offerMessage(channelSubsRef, channel, msg) *> releaseStream(channelSubsRef, channel)
+        case msg @ PushProtocol.Message(channel, _)    => offerMessage(channelSubsRef, channel, msg)
+        case msg @ PushProtocol.PSubscribe(pattern, _) => offerMessage(patternSubsRef, pattern, msg)
+        case msg @ PushProtocol.PUnsubscribe(pattern, _) =>
+          offerMessage(patternSubsRef, pattern, msg) *> releaseStream(patternSubsRef, pattern)
+        case msg @ PushProtocol.PMessage(pattern, _, _) => offerMessage(patternSubsRef, pattern, msg)
+      }
+  }
+
+  private def resubscribe: IO[RedisError, Unit] = {
+    def makeCommand(name: String, keys: Chunk[String]) =
+      if (keys.isEmpty)
+        Chunk.empty
+      else
+        RespValue
+          .Array((CommandNameInput.encode(name) ++ Input.Varargs(StringInput).encode(keys)).args.map(_.value))
+          .asBytes
+
+    for {
+      channels <- channelSubsRef.get.map(_.keys)
+      patterns <- patternSubsRef.get.map(_.keys)
+      commands = makeCommand(Subscription.Subscribe, Chunk.fromIterable(channels)) ++
+                   makeCommand(Subscription.PSubscribe, Chunk.fromIterable(patterns))
+      _ <- connection
+             .write(commands)
+             .when(commands.nonEmpty)
+             .mapError(RedisError.IOError(_))
+             .retryWhile(True)
+    } yield ()
+  }
+
+  /**
+   * Opens a connection to the server and launches receive operations. All failures are retried by opening a new
+   * connection. Only exits by interruption or defect.
+   */
+  val run: IO[RedisError, AnyVal] =
+    ZIO.logTrace(s"$this PubSub sender and reader has been started") *>
+      (send.repeat(Schedule.forever) race receive)
+        .tapError(e => ZIO.logWarning(s"Reconnecting due to error: $e") *> resubscribe)
+        .retryWhile(True)
+        .tapError(e => ZIO.logError(s"Executor exiting: $e"))
+}
+
+private[redis] object SingleNodeSubscriptionExecutor {
```

Comment:
This whole module if full of copy-paste. Let's try to reduce that.

Reply:
I made `SingleNodeRunner` trait that has two abstract functions `send`, `receive` to extract `run` function